### DISCUSSION TITLE
Trying to post an object with useFetch

### DISCUSSION THREAD

**Discussion by markteekman** [vueuse/vueuse#1305] [Q&A] [answered]
Hello everyone,

I'm trying to figure out how to post an object using `useFetch`. Now, this may also be a lack of JS skills, but I thought something like this would work:

```js
const {
  data = {
    username: values.email,
    password: values.password
  }
} = useFetch('https://mywebsite.com/wp-json/jwt-auth/v1/token').post().json()
```

But I get an error from the API saying the username is missing. For the record, this is the normal fetch request which works:

```js
const loginData = {
  username: values.email,
  password: values.password
}

fetch('https://mywebsite.com/wp-json/jwt-auth/v1/token', {
  method: 'POST',
  headers: {
    'Content-Type': 'application/json'
  },
  body: JSON.stringify(loginData)
})
  .then(response => response.json())
  .then(response => {
    console.log(response.data)
  })
  .catch((error) => {
    console.log(error.response.data)
  })
```

I really like the idea of how clean `useFetch` makes our code so I was hoping someone could help me understand it better in this  particular use case 🙂

**Comment by wheatjs** [comment, marked as answer]
Hmm not really sure how that code would work exactly 😅 

To post data using `useFetch` you will want to pass your data into the `post` method itself like this

```ts
const { data } = useFetch('https://mywebsite.com/wp-json/jwt-auth/v1/token').post({
    username: values.email,
    password: values.password
  })
  .json()
```

**Reply by markteekman** [reply]
Wow, thanks for the quick reply @wheatjs! That one actually makes more sense.. I'm still learning some core concepts in JavaScript so I get misguided sometimes 🥲 

Thanks for your help!

**Comment by markteekman** [comment]
@wheatjs I'm aware that this has probably more to do with Vue 3 and Refs, but do you know how I can access properties on the data object that returns when using `useFetch`?

![Schermafbeelding 2022-02-22 om 11 19 37](https://user-images.githubusercontent.com/3909046/155112180-bc12a8e8-25c5-4d87-b575-9f5f1355d75e.png)

I would like to access the token and would have expected that `console.log(data.value.data.token)` would yield it, but I'm getting an `undefined` on the `data.value`..

**Reply by wheatjs** [reply]
Sure there are a few ways you can go about it. The first is to watch the data object using Vue 3 `watch` function

```ts
import { watch } from 'vue'
import { useFetch } from '@vueuse/core'

const { data } = useFetch('https://mywebsite.com/wp-json/jwt-auth/v1/token').post({ username: values.email, password: values.password}).json()

watch(data, () => {
  if (data.value)
    console.log(data.value)
})
```

or you can use the [Suspense API](https://vuejs.org/guide/built-ins/suspense.html) and await the `useFetch`. If you go with this method you will have to wrap this component in a `<Suspense>` tag whenever you use it though

```ts
import { useFetch } from '@vueuse/core'

const { data } = await useFetch('https://mywebsite.com/wp-json/jwt-auth/v1/token').post({ username: values.email, password: values.password}).json()

console.log(data.value)
```

**Reply by markteekman** [reply]
Thanks a lot @wheatjs! I'm learning quite the things from your examples 🙂 Now, if I would want to show a message like 'Authenticating' I would be able to do that with `isFetching` right? I now have a function setup for when someone presses login after filling in their credentials, I would like to show the `isLoading` part while `isFetching === true`. It would look like this: 

```vue
<template>
  <p v-if="isLoading">Authenticating..</p>
  <VLogin
    @on-submit="handleLogin"
  />
</template>

<script setup>
import { ref, watch } from 'vue'
import { useFetch } from '@vueuse/core'

let isLoading = ref(false)

function handleLogin (values) {
const { data, isFetching, isFinished } = useFetch('https://mywebsite.com/wp-json/jwt-auth/v1/token').post({ username: values.email, password: values.password}).json()

  watch(data, () => {
    if (data.value) {
      console.log(data.value.data.token)
    }
  })
}
</script>
```

Then where in this function could I assign `isLoading` to `isFetching`, and then set `isLoading` to `isFinished` when it's done? Or should I rather use `isFetching` for the `v-if` statement?

**Reply by wheatjs** [reply]
Here is what I would suggest doing

```html
<template>
  <p v-if="isFetching">Authenticating..</p>
  <VLogin @on-submit="handleLogin" />
</template>

<script setup>
import { ref, watch } from 'vue'
import { useFetch } from '@vueuse/core'

const username = ref('')
const password = ref('')

const { data, isFetching, execute } = useFetch('https://mywebsite.com/wp-json/jwt-auth/v1/token', { immediate: false }).post({ username, password }).json()

watch(data, () => {
  if (data.value) {
    console.log(data.value.data.token)
  }
})

function handleLogin(values) {
  username.value = values.username
  password.value = values.password
  execute()
}
</script>
```

**Reply by markteekman** [reply]
@wheatjs thanks for explaining this in such detail, it makes much more sense now! And I really like how clean the code is in the end 😃 Thank you for taking the time to help me out 👍🏼

**Reply by wheatjs** [reply]
> @wheatjs thanks for explaining this in such detail, it makes much more sense now! And I really like how clean the code is in the end 😃 Thank you for taking the time to help me out 👍🏼 

Happy to help!

**Reply by Glandos** [reply]
> or you can use the [Suspense API](https://vuejs.org/guide/built-ins/suspense.html) and await the `useFetch`. If you go with this method you will have to wrap this component in a `<Suspense>` tag whenever you use it though
> 
> ```ts
> import { useFetch } from '@vueuse/core'
> 
> const { data } = await useFetch('https://mywebsite.com/wp-json/jwt-auth/v1/token').post({ username: values.email, password: values.password}).json()
> 
> console.log(data.value)
> ```

Typescript says me that `'await' has no effect on the type of this expression.ts(80007)` And indeed, my tests show me that at least in version 7.6 ~~the returned value is not a promise. It is only a promise with `useFetch` but not after a call to `post()`.~~ It work with `post` but not `delete`.

EDITED: works with post.
